File: Patterns/Facade Pattern/Cpp/HomeTheaterFacade.cpp
#pragma once

#include <iostream>
#include <string>

#include "./HomeTheater/Amplifier.cpp"
#include "./HomeTheater/CdPlayer.cpp"
#include "./HomeTheater/DvdPlayer.cpp"
#include "./HomeTheater/PopcornPopper.cpp"
#include "./HomeTheater/Projector.cpp"
#include "./HomeTheater/Screen.cpp"
#include "./HomeTheater/TheaterLights.cpp"
#include "./HomeTheater/Tuner.cpp"

using namespace std;

class HomeTheaterFacade {
private:
    Amplifier* amp;
    DvdPlayer* dvd;
    CdPlayer* cd;
    Projector* projector;
    PopcornPopper* popper;
    Screen* screen;
    TheaterLights* lights;
    Tuner* tuner;

public:
    HomeTheaterFacade(Amplifier* amp, DvdPlayer* dvd, CdPlayer* cd, Projector* projector, PopcornPopper* popper, Screen* screen, TheaterLights* lights, Tuner* tuner) :
        amp{amp}, dvd{dvd}, cd{cd}, projector{projector}, popper{popper}, screen{screen}, lights{lights}, tuner{tuner} {}
    
    void watchMovie(string movie) {
        cout << "Get ready to watch a movie...\n";
        popper->on();
        popper->pop();
        lights->dim(10);
        screen->down();
        projector->on();
        projector->wideScreenMode();
        amp->on();
        amp->setDvd(dvd);
        amp->setSurroundSound();
        amp->setVolumn(5);
        dvd->on();
        dvd->play(movie);
    }

    void endMovie() {
        cout << "Shutting movie theater down...\n";
        popper->off();
		lights->on();
		screen->up();
		projector->off();
		amp->off();
		dvd->stop();
		dvd->eject();
		dvd->off();
    }
};

File: Patterns/Facade Pattern/CSharp/HomeTheaterFacade.cs
using System;

using HomeTheater;

public class HomeTheaterFacade
{
    private HomeTheater.Amplifier amp;
    private HomeTheater.DvdPlayer dvd;
    private HomeTheater.CdPlayer cd;
    private HomeTheater.Projector projector;
    private HomeTheater.PopcornPopper popper;
    private HomeTheater.Screen screen;
    private HomeTheater.TheaterLights lights;
    private HomeTheater.Tuner tuner;

    public HomeTheaterFacade(HomeTheater.Amplifier amp, HomeTheater.DvdPlayer dvd, HomeTheater.CdPlayer cd, 
        Projector projector, PopcornPopper popper, Screen screen, TheaterLights lights, Tuner tuner) 
    {
        this.amp = amp;
        this.dvd = dvd;
        this.cd = cd;
        this.projector = projector;
        this.popper = popper;
        this.screen = screen;
        this.lights = lights;
        this.tuner = tuner;
    }

    public void WatchMovie(string movie)
    {
        Console.WriteLine("Get ready to watch a movie...");
        popper.On();
        popper.Pop();
        lights.Dim(10);
        screen.Down();
        projector.On();
        projector.WideScreenMode();
        amp.On();
        amp.SetDvd(dvd);
        amp.SetSurroundSound();
        amp.SetVolumn(5);
        dvd.On();
        dvd.Play(movie);
    }

    public void EndMovie() {
        Console.WriteLine("Shutting movie theater down...");
        popper.Off();
		lights.On();
		screen.Up();
		projector.Off();
		amp.Off();
		dvd.Stop();
		dvd.Eject();
		dvd.Off();
    }
}